common/src/main/java/io/pravega/common/util/SimpleDeque.java

```diff
@@ -0,0 +1,315 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.base.Preconditions;
+import java.util.AbstractCollection;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Queue;
+import javax.annotation.concurrent.NotThreadSafe;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Simplified {@link java.util.Deque} implementation that provides an efficient method to remove multiple items at once
+ * from the head.
+ *
+ * Note that this class does not implement {@link java.util.Deque} and as such not all methods in that interface are
+ * present here. The only reason it exists is because there is no way in {@link java.util.ArrayDeque} to efficiently
+ * remove multiple items at once (see {@link #pollFirst(int)}.
+ *
+ * @param <T> Type of item in the {@link SimpleDeque}.
+ */
+@NotThreadSafe
+public class SimpleDeque<T> {
+    //region Members
+
+    private static final int MAX_CAPACITY = Integer.MAX_VALUE - Long.BYTES;
+    private static final int MAX_HALF = MAX_CAPACITY / 2;
+    private Object[] items;
+    private int head;
+    private int tail;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the {@link SimpleDeque} class.
+     */
+    public SimpleDeque() {
+        this(16);
```

Comment:
This class is very similar to Java's `ArrayDeque`. That class also chooses 16 as a default init value.

For our purposes (`BlockingDrainingQueue`) this initial value makes little difference. We only create the queue once and then hang on to it. So it will eventually grow to a "steady-state" length and then reuse the same element array.